### ISSUE TITLE
harden(inference): CandidateSet::argmax returns out-of-set token 0 for all-(-inf) set (Metal+CPU reachable)

### ISSUE THREAD

**Issue by ohdearquant** [ohdearquant/lattice#405]
## Summary

`CandidateSet::argmax` (`crates/inference/src/sampling.rs:~94`) returns a **hardcoded token id 0** when every candidate has logit `NEG_INFINITY` (an all-masked / all-(-inf) compact set). Token 0 may **not be a member of the candidate set** after grammar-mask compaction or `retain_top_k`, so the sampler can emit an out-of-set token.

```rust
pub fn argmax(&self) -> u32 {
    let mut best_id = 0u32;                 // <-- returned as-is if nothing beats NEG_INFINITY
    let mut best_val = f32::NEG_INFINITY;
    for c in &self.candidates {
        if c.logit > best_val { best_val = c.logit; best_id = c.token_id; }
    }
    best_id
}
```

For an all-(-inf) set, no candidate satisfies `c.logit > NEG_INFINITY`, so `best_id` stays `0` regardless of which token ids are actually in the set.

## Reachability (live, not test-only)

`argmax` is called on real generation paths:
- **Live CPU**: `CandidateSet::apply_temperature` (sampling.rs:121) routes a degenerate temperature to `argmax` (the #399 one-hot path). An all-masked set reaching a degenerate-temperature draw returns token 0.
- **Metal**: `forward/metal_qwen35.rs:12898` and `:12940` (degenerate-temperature compact-set paths) call `cs.argmax()` directly.

An all-(-inf) `CandidateSet` is producible by a grammar mask that masks every surviving candidate, or by upstream poisoning. The canonical `Sampler` softmax path fails closed on all-(-inf) (per the softmax fail-closed class — `InvalidInput`), but `argmax` silently returns token 0 instead.

## Why this needs a decision

The fix requires **deciding the semantics of argmax over an all-masked set** — undefined behavior being decided for the first time, and the surface is Metal/GPU-reachable. Candidate decisions:

1. Return the **first candidate's** token id (`self.candidates.first().map(|c| c.token_id).unwrap_or(0)`) — guarantees an in-set token; minimal.
2. Return `Option<u32>` / `Result` and have callers fail closed (`InvalidInput`) — matches the softmax-path contract but changes the signature and all call sites (Metal included).
3. Treat all-masked as a hard error at the mask layer before argmax is reached.

Option 1 is the smallest correct change but still decides "argmax of an all-masked set = first member," and touches the Metal sampling path. Surfacing for a decision on the contract + the Metal-path implications.

## Recommendation

Option 1 (init `best_id` to the first candidate's token id) as the minimal correctness fix — it can never return an out-of-set token and is behavior-identical for any set with ≥1 finite logit. Escalated rather than self-merged because it (a) decides previously-undefined all-masked semantics and (b) is reachable from the Metal sampler.

## Provenance

Found by a discovery review pass over `crates/inference/src/sampling.rs` selection paths. Companion finding (qwen35 live-sampler NaN policy) filed separately. Both need a decision; neither self-merged. Relates to the softmax fail-closed bug class and the shared-guard-public-boundary pattern (a guard on the softmax path does not protect the `argmax` building block).
